### DISCUSSION TITLE
Support dynamic loading of scripts and gamemodes

### DISCUSSION THREAD

**Discussion by samuelallan72** [piqueserver/piqueserver#653] [Feature Requests]
This would be helpful for faster debugging/testing of scripts. May or may not be complex to implement.

**Comment by NotAFile** [comment]
With the current system it is borderline impossible... :/

Since the scripts all subclass the main class, you can't really disable a script without deleting the core `protocol` obect, which contains most of the servers state.

**Comment by NotAFile** [comment]
I thought I'd just elaborate on my plans here: I'd like to add shim classes that pretend to be Protocol and Connection objects, but expose a well-defined API instead. These would then be passed to the apply_scripts function. This means that plugins won't be able to do everything anymore, we have to manually place hooks in the places where we want plugins to hook into, but it does allow us to get rid of the subclassing nightmare internally without breaking compatibility.
